java/lib.nbjavac/src/org/netbeans/lib/nbjavac/CheckProcessor.java

```diff
@@ -0,0 +1,88 @@
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.netbeans.lib.nbjavac;
+
+import com.sun.source.tree.MethodInvocationTree;
+import com.sun.source.util.JavacTask;
+import com.sun.source.util.TaskEvent;
+import com.sun.source.util.TaskListener;
+import com.sun.source.util.TreePath;
+import com.sun.source.util.TreePathScanner;
+import com.sun.source.util.Trees;
+import java.util.Set;
+import javax.annotation.processing.AbstractProcessor;
+import javax.annotation.processing.RoundEnvironment;
+import javax.annotation.processing.SupportedAnnotationTypes;
+import javax.lang.model.SourceVersion;
+import javax.lang.model.element.Element;
+import javax.lang.model.element.ElementKind;
+import javax.lang.model.element.QualifiedNameable;
+import javax.lang.model.element.TypeElement;
+import javax.tools.Diagnostic;
+
+@SupportedAnnotationTypes("*")
+public class CheckProcessor extends AbstractProcessor implements TaskListener {
+
+    private boolean listenerInstalled;
+
+    @Override
+    public boolean process(Set<? extends TypeElement> annotations, RoundEnvironment roundEnv) {
+        if (!listenerInstalled) {
+            JavacTask.instance(processingEnv).addTaskListener(this);
+            listenerInstalled = true;
+        }
+
+        return false;
+    }
+
+    @Override
+    public void finished(TaskEvent e) {
+        if (e.getKind() != TaskEvent.Kind.ANALYZE) {
+            return ;
+        }
+        String path = e.getCompilationUnit().getSourceFile().toUri().toString();
+        if (path.contains("test") || !path.contains("org/netbeans")) {
+            //only applies to NetBeans code, and tests are excepted
+            return ;
+        }
+        new TreePathScanner<Void, Void>() {
+            @Override
+            public Void visitMethodInvocation(MethodInvocationTree node, Void p) {
+                TreePath selectPath = new TreePath(getCurrentPath(), node.getMethodSelect());
+                Trees trees = Trees.instance(processingEnv);
+                Element el = trees.getElement(selectPath);
+
+                if (el != null &&
+                    el.getKind() == ElementKind.METHOD &&
+                    ((QualifiedNameable) el.getEnclosingElement()).getQualifiedName().contentEquals("org.netbeans.api.java.source.TreeMaker") &&
+                    (el.getSimpleName().contentEquals("Class") || el.getSimpleName().contentEquals("Interface")) &&
+                    processingEnv.getElementUtils().isDeprecated(el)) {
+                    trees.printMessage(Diagnostic.Kind.ERROR, "Use of the deprecated TreeMaker.Class/Interface method is not permitted inside the NetBeans sources.", node, selectPath.getCompilationUnit());
```

Comment:
this is interesting. Will this run while nb-javac is used within the NB editor too or only during build?

I suppose this is picked up when `-proc:full` is set + module dependency to nb-javac lib is declared?

Reply:
Sorry for belated reply.

(AFAIK) This will run every time this module is on the appropriate path (and the AP is enabled). I.e. it will run in the editor when the NetBeans modules that depend on this are open (like `java.source.base`). It will not (should not) run for ordinary unrelated user code.

Reply:
I have nothing against it but I am just wondering: given that the methods are deprecated and all call sites were updated in this PR (it is also only checking `org/netbeans` paths and nothing external), I am wondering what this is supposed to catch?

Wouldn't grepping the build for compiler warnings related to `TreeMaker` be sufficient?

Reply:
Sure, grepping for compiler warnings would work. The idea is mostly to make sure we don't re-introduce invocations of the deprecated methods - there are surely various means to achieve that, so it is mostly because it is easier for me to write an AP than a filter that would check for the specific warnings during a build.

Reply:
can hint files mark things as error? NB used to use those for migration purposes, e.g:
https://github.com/apache/netbeans/blob/master/platform/openide.filesystems/src/META-INF/upgrade/FileObject.hint

I guess I am wondering why a new way of API usage warnings is introduced now - but maybe I am missing something and the other approaches would not work here. I think it is also fairly unlikely that new code would start calling a deprecated method by accident - but i could be wrong.

Reply:
```java
"don't use this":
new Character($l) :: $l instanceof char =>
<!error='blabla'>
;;
```

tried this but it looks like that it won't mark it as error unfortunately. But it marks it as regular warning.

the only usage for the error option is here https://github.com/apache/netbeans/blob/3ebd6cb396842f212bd09011d689941884dd649d/java/java.hints.declarative/src/org/netbeans/modules/java/hints/declarative/DeclarativeHintsWorker.java#L135

Reply:
thanks, but i was only asking questions - I was curious why it was implemented that way thats all.